src/actors/objects/mod.rs

```diff
@@ -552,29 +596,44 @@ impl ObjectsActor {
         file_metas.and_then(move |responses: Vec<Result<Arc<ObjectFileMeta>, _>>| {
             responses
                 .into_iter()
-                .enumerate()
-                .min_by_key(|(i, response)| {
+                .min_by_key(|response| {
                     // Prefer files that contain an object over unparseable files
-                    let object = match response {
-                        Ok(object) => object,
+                    let object_meta = match response {
+                        Ok(object_meta) => object_meta,
                         Err(e) => {
                             log::debug!("Failed to download: {:#?}", e);
-                            return (3, *i);
+                            return 3;
                         }
                     };
 
                     // Prefer object files with debug/unwind info over object files without
-                    let score = match purpose {
-                        ObjectPurpose::Unwind if object.features.has_unwind_info => 0,
-                        ObjectPurpose::Debug if object.features.has_debug_info => 0,
-                        ObjectPurpose::Debug if object.features.has_symbols => 1,
-                        ObjectPurpose::Source if object.features.has_sources => 0,
+                    match purpose {
+                        ObjectPurpose::Unwind if object_meta.features.has_unwind_info => 0,
+                        ObjectPurpose::Debug if object_meta.features.has_debug_info => 0,
+                        ObjectPurpose::Debug if object_meta.features.has_symbols => 1,
+                        ObjectPurpose::Source if object_meta.features.has_sources => 0,
                         _ => 2,
-                    };
-
-                    (score, *i)
+                    }
+                })
+                .filter(|response| match response {
+                    // Make sure we only return objects which provide the requested info
+                    Ok(object_meta) => {
+                        if object_meta.status == CacheStatus::Positive {
+                            // object_meta.features is meaningless when CacheStatus != Positive
+                            match purpose {
+                                ObjectPurpose::Unwind => object_meta.features.has_unwind_info,
+                                ObjectPurpose::Debug => {
+                                    object_meta.features.has_debug_info
+                                        || object_meta.features.has_symbols
+                                }
+                                ObjectPurpose::Source => object_meta.features.has_sources,
+                            }
+                        } else {
+                            true
+                        }
+                    }
+                    Err(_) => true,
```

Comment:
just out of curiosity: the new filter now throws away things we can’t use, but why does it pass through errors and negative cache statuses?

Reply:
because all those could have been object files which might have contained the info we needed.  but we don't know until we can actually fetch it.  and passing them through allows us to report on their status later (download error, malformed, negative cache).  this whole behaviour of only returning one result here a bit questionable and may have to change to improve reporting